federation/pkg/federation-controller/ingress/ingress_controller.go

```diff
@@ -0,0 +1,325 @@
+			return err
+		},
+		func(client federation_release_1_4.Interface, obj pkg_runtime.Object) error {
+			ingress := obj.(*extensions_v1beta1.Ingress)
+			err := client.Extensions().Ingresses(api.NamespaceAll).Delete(ingress.Name, &api.DeleteOptions{})
+			return err
+		})
+	return ic
+}
+
+func (ic *IngressController) Run(stopChan <-chan struct{}) {
+	glog.Infof("Starting Ingress Controller")
+	go ic.ingressInformerController.Run(stopChan)
+	glog.Infof("... Starting Ingress Federated Informer")
+	ic.ingressFederatedInformer.Start()
+	go func() {
+		<-stopChan
+		glog.Infof("Stopping Ingress Controller")
+		ic.ingressFederatedInformer.Stop()
+	}()
+	ic.ingressDeliverer.StartWithHandler(func(item *util.DelayingDelivererItem) {
+		i := item.Value.(string)
+		glog.V(4).Infof("Ingress change delivered, reconciling: %v", i)
+		ic.reconcileIngress(i)
+	})
+	ic.clusterDeliverer.StartWithHandler(func(_ *util.DelayingDelivererItem) {
+		glog.V(4).Infof("Cluster change delivered, reconciling ingresses")
+		ic.reconcileIngressesOnClusterChange()
+	})
+	go func() {
+		select {
+		case <-time.After(time.Minute):
+			glog.V(4).Infof("Ingress controller is garbage collecting")
+			ic.ingressBackoff.GC()
+		case <-stopChan:
+			return
+		}
+	}()
+}
+
+func (ic *IngressController) deliverIngressObj(obj interface{}, delay time.Duration, failed bool) {
+	ingress := obj.(*extensions_v1beta1.Ingress)
+	ic.deliverIngress(ingress.Name, delay, failed)
+}
+
+func (ic *IngressController) deliverIngress(ingress string, delay time.Duration, failed bool) {
+	glog.V(4).Infof("Delivering ingress: %q", ingress)
+	if failed {
+		ic.ingressBackoff.Next(ingress, time.Now())
+	} else {
+		ic.ingressBackoff.Reset(ingress)
+	}
+	ic.ingressDeliverer.DeliverAfter(ingress, ingress, delay)
+}
+
+// Check whether all data stores are in sync. False is returned if any of the informer/stores is not yet
+// synced with the coresponding api server.
+func (ic *IngressController) isSynced() bool {
+	if !ic.ingressFederatedInformer.ClustersSynced() {
+		glog.V(2).Infof("Cluster list not synced")
+		return false
+	}
+	clusters, err := ic.ingressFederatedInformer.GetReadyClusters()
+	if err != nil {
+		glog.Errorf("Failed to get ready clusters: %v", err)
+		return false
+	}
+	if !ic.ingressFederatedInformer.GetTargetStore().ClustersSynced(clusters) {
+		glog.V(2).Infof("Target store not synced")
+		return false
+	}
+	glog.V(4).Infof("Cluster list is synced")
+	return true
+}
+
+// The function triggers reconcilation of all federated ingresses.
+func (ic *IngressController) reconcileIngressesOnClusterChange() {
+	glog.V(4).Infof("Reconciling ingresses on cluster change")
+	if !ic.isSynced() {
+		ic.clusterDeliverer.DeliverAt(allClustersKey, nil, time.Now().Add(ic.clusterAvailableDelay))
+	}
+	for _, obj := range ic.ingressInformerStore.List() {
+		ingress := obj.(*extensions_v1beta1.Ingress)
+		ic.deliverIngress(ingress.Name, ic.smallDelay, false)
+	}
+}
+
+func backoff(trial int64) time.Duration {
```

Comment:
backoff is now handled with util/lifecycle/backoff. please take a look at namespace controller.